book/src/dev/rfcs/0003-state-updates.md

```diff
@@ -0,0 +1,305 @@
+be specified in a later RFC. If contextual validation checks succeed, the new
+block is added to one of the in-memory chains. If the resulting chain is
+longer than 100 blocks, the oldest block is now past the reorg limit, so we
+remove it from all in-memory chains and commit it to sled as described below
+in `CommitFinalizedBlock`.
+
+Finally, process any queued children of the newly committed block the same way.
+
+### `Request::CommitFinalizedBlock(Arc<Block>)`
+[request-commit-finalized-block]: #request-finalized-block
+
+Commits a finalized block to the sled state, skipping contextual validation.
+The block's parent must be the current sled tip. This is exposed for use in
+checkpointing, which produces in-order finalized blocks. Returns
+`Response::Added(BlockHeaderHash)` with the hash of the committed block if
+successful.
+
+This should be implemented as a wrapper around a function also called by
+[`Request::CommitBlock`](#request-commit-block), which should:
+
+1. Obtain the highest entry of `hash_by_height` as `(old_height, old_tip)`.
+Check that `block`'s parent hash is `old_tip` and its height is
+`old_height+1`, or panic.  This check is performed as defense-in-depth
+to prevent database corruption, but it is the caller's responsibility to
+commit finalized blocks in order.
+
+2. Insert `(block_hash, block)` into `blocks_by_hash` and
+   `(BE32(height), block_hash)` into `hash_by_height`.
+
+3. Iterate over the enumerated transactions in the block. For each transaction:
+
+   1. Insert `(transaction_hash, block_hash || BE32(tx_index))` to
+   `tx_by_hash`;
+
+   2. For each `TransparentInput::PrevOut { outpoint, .. }` in the
+   transaction's `inputs()`, remove `outpoint` from `utxo_by_output`.
+
+   3. For each `output` in the transaction's `outputs()`, construct the
+   `outpoint` that identifies it, and insert `(outpoint, output)` into
+   `utxo_by_output`.
+
+   4. For each [`JoinSplit`] description in the transaction,
+   insert `(nullifiers[0],())` and `(nullifiers[1],())` into
+   `sprout_nullifiers`.
+
+   5. For each [`Spend`] description in the transaction, insert
+   `(nullifier,())` into `sapling_nullifiers`.
+
+[`JoinSplit`]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.JoinSplit.html
+[`Spend`]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.Spend.html
+
+    
+These updates can be performed in a batch or without necessarily iterating
+over all transactions, if the data is available by other means; they're
+specified this way for clarity.
+
+### `Request::Depth(BlockHeaderHash)`
+[request-depth]: #request-depth
+
+Computes the depth in the best chain of the block identified by the given
+hash, returning
+
+- `Response::Depth(Some(depth))` if the block is in the main chain;
+- `Response::Depth(None)` otherwise.
+
+### `Request::Tip`
+[request-tip]: #request-tip
+
+Returns `Response::Tip(BlockHeaderHash)` with the current best chain tip.
+
+### `Request::BlockLocator`
```

Comment:
We pass the genesis block, so it can be used as the default locator if the state is empty.

```suggestion
### `Request::BlockLocator(genesis: block::Hash)`
```

Reply:
We currently do this -- should we continue to do it in the future? I think this is more awkward than we need it to be. Can't the state service return the genesis hash on its own if it's empty?

Reply:
I think the original design was an artefact of the early code structure, where only the sync service knew about the current network.

I opened #1024 to make this change.